### DISCUSSION TITLE
Connection issues between Crazyradio 2.0 & Crazyflie 2.1+

### DISCUSSION THREAD

**Discussion by judicaelclair** [bitcraze/discussions#1682] [Q&A]
**UPDATE:** Downgrading Crazyradio 2.0 firmware from 1.2 down to 1.1 fixes all connectivity issues mentioned below.

I recently acquired a pair of Crazyflie 2.1+ (CF). With the minimal setup (i.e. no decks), I'm having a lot of trouble connecting to the CFs using a Crazyradio 2.0 (CR). I'm using the latest firmware releases for both the CF and CR. When running on a Ubuntu 22.04.5 machine, to successfully connect to a CF, I must:
1) Switch on CF.
2) Plug in CR.
3) Scan for CF (if using cfclient).
4) **Unplug CR then plug it back in.**
5) Connect to CF.

That is, **before every connection to a CF, I need to unplug the CR at least once**. This is because after CR disconnects, the CR may:
1) Stop being detected.
2) Start timing out when scanning.
3) Stall whilst establishing a new connection to a CF.

Example Errors:
```
INFO:cflib.crtp.radiodriver:v3.2 dongle with serial N/A found
===== 2024.12.06 21:12:20 =====
Exception in thread Shared Radio:
Traceback (most recent call last):
  File "lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "lib/python3.10/site-packages/cflib/crtp/radiodriver.py", line 207, in run
    resp = self._radio.scan_channels(start, stop, packet)
  File "lib/python3.10/site-packages/cflib/drivers/crazyradio.py", line 272, in scan_channels
    self.set_channel(i)
  File "lib/python3.10/site-packages/cflib/drivers/crazyradio.py", line 184, in set_channel
    _send_vendor_setup(self.handle, SET_RADIO_CHANNEL, channel, 0, ())
  File "lib/python3.10/site-packages/cflib/drivers/crazyradio.py", line 323, in _send_vendor_setup
    handle.ctrl_transfer(usb.TYPE_VENDOR, request, wValue=value,
  File "lib/python3.10/site-packages/usb/core.py", line 1082, in ctrl_transfer
    ret = self._ctx.backend.ctrl_transfer(
  File "lib/python3.10/site-packages/usb/backend/libusb1.py", line 893, in ctrl_transfer
    ret = _check(self.lib.libusb_control_transfer(
  File "lib/python3.10/site-packages/usb/backend/libusb1.py", line 602, in _check
    raise USBTimeoutError(_strerror(ret), ret, _libusb_errno[ret])
usb.core.USBTimeoutError: [Errno 110] Operation timed out
```
```
Cannot find a Crazyradio Dongle
INFO:cflib.drivers.cfusb:Looking for devices....
INFO:cflib.crtp.prrtdriver:Initialized PRRT driver.
```
```
[Errno 2] Entity not found
INFO:cflib.drivers.cfusb:Looking for devices....
INFO:cflib.crtp.prrtdriver:Initialized PRRT driver.
```
Endless printing of:
```
INFO:cflib.crtp.radiodriver:Dongle reported ACK status == None
```

Note that:
- I get similar behaviour/issues on a Windows 11 machine, but I only need things working on Ubuntu.
- I replicated the issues on two different Ubuntu 22.04.5 machines.
- I have two crazyflies and two crazyradios. All combinations exhibit same issues.
- I've tried different channels: 40, 80.
- I've tried different data rates: 1M, 2M.

**Comment by ataffanel** [comment]
Thanks a lot for the detailed report! It is very strange that the new version of the radio firmware would have such a big effect. I am going to investigate and hoppefully fix the issue in a new release.

**Comment by reucsan** [comment]
I'd just like to confirm @judicaelclair 's findings, I have the same issue:
`INFO:cflib.crtp.radiodriver:Dongle reported ACK status == None` 
Downgrading to 1.1 fixes it for me too.

**Comment by ataffanel** [comment]
I have already retracted version 1.2 since it is indeed crashing. I have now created an issue so that we can track the bug and the resolution: https://github.com/bitcraze/crazyradio2-firmware/issues/8
